p2p/router.go

```diff
@@ -774,12 +716,38 @@ func (r *Router) handshakePeer(ctx context.Context, conn Connection, expectID No
 	return peerInfo, peerKey, nil
 }
 
+func (r *Router) runWithPeerLock(setup func() error) error {
+	r.peerMtx.Lock()
+	defer r.peerMtx.Lock()
+	return setup()
+}
+
 // routePeer routes inbound and outbound messages between a peer and the reactor
 // channels. It will close the given connection and send queue when done, or if
 // they are closed elsewhere it will cause this method to shut down and return.
-func (r *Router) routePeer(peerID NodeID, conn Connection, sendQueue queue) {
+func (r *Router) routePeer(peerID NodeID, conn Connection, mgrsetup func() error) {
+	if err := r.runWithPeerLock(mgrsetup); err != nil {
+		r.logger.Error("failed registering peer connection, not routing", "err", err)
+		return
+	}
```

Comment:
Does it make sense adding this callback function - I think it affects readability - Especially as it's at the start of the function. Why not just add that logic before `routePeer` is called

Reply:
I agree here.

Reply:
So the crux of this PR is that we have to make the calls to the peer manager inside of the router's peer mutex. Otherwise you can end up in situations where incoming connections and outgoing connections can "race" with the effect of both connections ending up closed. 

So I can avoid passing the setup function object into route peer, but it ends up making routePeer less reliable, and the replacement is just to put a function object that we call in a lock before routePeer, so I think the code, as written is actually more clear/safe, but I'm willing to go either way.